packages/eslint-plugin/src/rules/restrict-template-expressions.ts

```diff
@@ -12,16 +12,35 @@ import {
   isTypeNeverType,
 } from '../util';
 
-type Options = [
-  {
-    allowAny?: boolean;
-    allowBoolean?: boolean;
-    allowNullish?: boolean;
-    allowNumber?: boolean;
-    allowRegExp?: boolean;
-    allowNever?: boolean;
+const optionsObj = {
+  Any: isTypeAnyType,
+  Array: (type, checker, rec): boolean => {
+    if (!checker.isArrayType(type)) {
+      return false;
+    }
+    const numberIndexType = type.getNumberIndexType();
+    return !numberIndexType || rec(numberIndexType);
   },
-];
+  Boolean: (type): boolean => isTypeFlagSet(type, ts.TypeFlags.BooleanLike),
+  Nullish: (type): boolean =>
+    isTypeFlagSet(type, ts.TypeFlags.Null | ts.TypeFlags.Undefined),
+  Number: (type): boolean =>
+    isTypeFlagSet(type, ts.TypeFlags.NumberLike | ts.TypeFlags.BigIntLike),
+  RegExp: (type, checker): boolean => getTypeName(checker, type) === 'RegExp',
+  Never: isTypeNeverType,
+} satisfies Record<
+  string,
+  (
+    type: ts.Type,
+    checker: ts.TypeChecker,
+    rec: (type: ts.Type) => boolean,
+  ) => boolean
+>;
```

Comment:
[Naming] `fn` and `rec` aren't super informative names. I had a hard time figuring out what this does without scanning back and forth through other code. Maybe... `recurser`? `typeTester`? Names like that?

Extracting out a `type OptionTypeTester` for the `(type, checker, rec)` functions also helped me read it a bit.